File: src/main/java/com/fullstack/demo/entity/Destination.java
package com.fullstack.demo.entity;

import com.fullstack.demo.entity.DailyPlan;
import jakarta.persistence.*;
import lombok.*;
import com.fullstack.demo.entity.DestinationType;

// Destination.java
@Entity
@Getter
@Setter
@Builder
@AllArgsConstructor
@NoArgsConstructor
public class Destination {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long destinationId;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private DestinationType type;

    @Column(nullable = false)
    private String name;

    @Column(nullable = false)
    private String address;

    @Column(nullable = false)
    private double latitude;

    @Column(nullable = false)
    private double longitude;

    @Column(length = 1000)
    private String transitDetails;  // 대중교통 이용시 이 장소로 오는 경로

    @Column(length = 1000)
    private String lastTransitDetails;  // 대중교통 이용시 이 장소에서 숙소로 가는 경로

    @Column(length = 1000)
    private String drivingDetails;  // 자동차 이용시 이 장소로 오는 경로

    @Column(length = 1000)
    private String lastDrivingDetails;  // 자동차 이용시 이 장소에서 숙소로 가는 경로

    private Integer orderInDay;
    private Double distanceToNext;
    private Integer timeToNext;

    @ManyToOne
    @JoinColumn(name = "daily_plan_id")
    private DailyPlan dailyPlan;

    @Builder
    public Destination(String name, String address, Double latitude, Double longitude, DestinationType type) {
        this.name = name;
        this.address = address;
        this.latitude = latitude;
        this.longitude = longitude;
        this.type = type;
    }


    public void setDailyPlan(DailyPlan dailyPlan) {
        this.dailyPlan = dailyPlan;
    }

    public void updateOrderAndDistanceInfo(Integer order, Double distanceToNext, Integer timeToNext) {
        this.orderInDay = order;
        this.distanceToNext = distanceToNext;
        this.timeToNext = timeToNext;
    }

    public static Destination createDestination(String name, String address, double latitude, double longitude, DestinationType type) {
        Destination destination = new Destination();
        destination.setName(name);
        destination.setAddress(address);
        destination.setLatitude(latitude);
        destination.setLongitude(longitude);
        destination.setType(type);
        return destination;
    }
}

File: lib/entity/Destination.dart
import 'dart:ffi';

class Destination {
  final int id;
  final DestinationType type;
  final String name;
  final String address;
  final double latitude;
  final double longitude;
  final String? transitDetails;
  final String? lastTransitDetails;
  final String? drivingDetails;
  final String? lastDrivingDetails;
  final int? orderInDay;
  final double? distanceToNext;
  final int? timeToNext;

  Destination({
    required this.id,
    required this.type,
    required this.name,
    required this.address,
    required this.latitude,
    required this.longitude,
    this.transitDetails,
    this.lastTransitDetails,
    this.drivingDetails,
    this.lastDrivingDetails,
    this.orderInDay,
    this.distanceToNext,
    this.timeToNext,
  });

  factory Destination.fromJson(Map<String, dynamic> json) {
    return Destination(
      id: json['destinationId'] ?? 0,  // null이면 0으로 처리
      type: DestinationType.values.firstWhere(
              (e) => e.toString() == 'DestinationType.${json['type']}'
      ),
      name: json['name'],
      address: json['address'],
      latitude: json['latitude'],
      longitude: json['longitude'],
      transitDetails: json['transitDetails'],
      lastTransitDetails: json['lastTransitDetails'],
      drivingDetails: json['drivingDetails'],
      lastDrivingDetails: json['lastDrivingDetails'],
      orderInDay: json['orderInDay'],
      distanceToNext: json['distanceToNext'],
      timeToNext: json['timeToNext'],
    );
  }

  Map<String, dynamic> toJson() => {
    'destinationId': id,
    'type': type.toString().split('.').last,
    'name': name,
    'address': address,
    'latitude': latitude,
    'longitude': longitude,
    'transitDetails': transitDetails,
    'lastTransitDetails': lastTransitDetails,
    'drivingDetails': drivingDetails,
    'lastDrivingDetails': lastDrivingDetails,
    'orderInDay': orderInDay,
    'distanceToNext': distanceToNext,
    'timeToNext': timeToNext,
  };
}

enum DestinationType {
  ACCOMMODATION,
  SPOT,
}